ADA/databricks_rstudio_sql_warehouse.qmd

```diff
@@ -1,105 +1,210 @@
+
+---
+
+## Pre-requisites
+
+Before you start, you must have:
+
+-   Access to the Databricks platform
 
--   Access to Databricks
--   Access to a SQL Warehouse on DataBricks
--   R and RStudio downloaded
+-   Access to data in a SQL Warehouse on Databricks
+
+-   R and RStudio downloaded and installed
+
+-   The `ODBC` and `DBI` packages installed in RStudio
+
+If you do not have access to Databricks or a SQL Warehouse within Databricks, you can request this using [a service request form](https://dfe.service-now.com/serviceportal?id=sc_cat_item&sys_id=74bc3be81b212d504f999978b04bcb0b).
+
+If you do not have R or RStudio, you can find them both in the Software Centre. Note that you need both R **and** RStudio installed.
+
+---
+
+## Process
+
+There are three steps to complete before your connection can be established. These are:
+
+-   Installing an ODBC driver on your laptop to enable a connection between your laptop and Databricks
+-   Modifying your .Renviron file to establish a connection between RStudio and Databricks
+-   Adding connection code to your existing scripts in RStudio
+
+Each of these steps is described in more detail in the sections below. 
+
+---
+
+### Setting up the ODBC driver
+
+An ODBC driver is required for the `ODBC` package in R to work - you must install it before attempting to use the package to connect to your data. 
+
+---
+
+#### Install the Simba Spark ODBC driver from the Software Centre
+
+---
+
+-   Open the Software Centre via the start menu
+
+-   In the 'Applications' tab, click `Simba Spark ODBC Driver 64-bit`
+
+::: {align="center"}
+![](../images/databricks-software-centre.png)
+:::
+
+-   Click install
 
 ------------------------------------------------------------------------
 
-# Downloading an ODBC driver
+### Establishing an RStudio connection using environment variables
 
 ------------------------------------------------------------------------
 
-1.  Install the 'Simba Spark ODBC' driver from the software centre.
+The `ODBC` package in RStudio allows you to connect to Databricks by creating and modifying four environment variables in your .Renviron file.
 
-    i)  Open the Software Centre via the start menu.
+To set the environment variables, call `usethis::edit_r_environ()`. You will then need to enter the following information:
 
-    ii) In the 'Applications' tab, click 'Simba Spark ODBC Driver 64-bit'. ![](../images/databricks-software-centre.png)
+```         
+DATABRICKS_HOST=<databricks-host>
+DATABRICKS_SQL_PATH=<sql-warehouse-path>
+DATABRICKS_TOKEN=<personal-access-token> 
+DATABRICKS_CLUSTER_ID=<personal-cluster-id>
```

Comment:
The DATABRICKS_CLUSTER_ID should also be the personal cluster path again rather than the ID.

Reply:
Good shout, I had it as ID before because I thought it only accepted that, but actually `sparklyr::spark_connect` accepts both `cluster_id` and `cluster_path`. Agree for consistency should be the same for both, and the connection string becomes:

```r
sc <- spark_connect(
  cluster_path = Sys.getenv("DATABRICKS_CLUSTER_PATH"),
  method = "databricks_connect"
)
```

where `DATABRICKS_CLUSTER_PATH=sql/protocolv1/o/12345...../abcde123....`

Reply:
To clarify this is what should be in the other guidance with using sparklyr, so this guidance is compatible with both sparklyr and odbc setup. For ODBC this would be either:

```r
con <- DBI::dbConnect(
  odbc::databricks(),
  httpPath = Sys.getenv("DATABRICKS_SQL_WAREHOUSE_PATH")
)
```
or
```r
con <- DBI::dbConnect(
  odbc::databricks(),
  httpPath = Sys.getenv("DATABRICKS_CLUSTER_PATH")
)
```

Reply:
Agreed on both points. 👍

Reply:
Correction on this, I just tested sparklyr in a clean environment and `cluster_path` doesn't work, it _must_ be `cluster_id`. This means we either:
1. Tell people to add `DATABRICKS_CLUSTER_ID` at the start of the `sparklyr` guidance.
2. Do some ugly regex on `CLUSTER_PATH` to get the bit after the last backslash:
```r
cluster_id = str_match(Sys.getenv("DATABRICKS_CLUSTER_PATH"),  ".*/(.*)$")[2]
```

I think option 1 is cleaner, and we tell people to define `DATABRICKS_CLUSTER_ID` and `DATABRICKS_CLUSTER_PATH` -- most people won't be using sparklyr so it won't matter too much.

Reply:
Agreed that option 1 is cleaner. I don't think there's anything wrong with having an additional environmental variable even if it is a little repetitive.